app/callbacks/views/sns.py

```diff
@@ -0,0 +1,416 @@
+    else:
+        current_app.logger.info(
+            "SubscriptionConfirmation succeeded for subscription {subscription_arn}",
+            extra={
+                "subscription_arn": subscription_arn,
+                "confirmation_request_id": confirmation_request_id,
+            },
+        )
+
+    return jsonify(status='ok'), 200
+
+
+@lru_cache()
+def _get_certificate(url):
+    with logged_duration(
+        logger=current_app.logger,
+        message=lambda _: (
+            "Fetched certificate for SNS signature validation from {target_url}"
+            if sys.exc_info()[0] is None else
+            # need to literally format() the exception into the message as it's difficult to get it injected into extra
+            "Failed to fetch certificate for SNS signature validation from {{target_url}}: {!r}".format(
+                sys.exc_info()[1]
+            )
+        ),
+        log_level=logging.INFO,
+        condition=True,
+    ) as log_context:
+        log_context.update({"target_url": url})
+        response = requests.get(url)
+        response.raise_for_status()
+        return response.content
+
+
+# deliberately omitting the optional trailing .cn here
+VALIDATION_CERTIFICATE_URL_REGEX = r"^https://[-a-z0-9.]+\.amazonaws\.com/"
+# give request extra headroom of a minute here
+VALIDATION_MAX_AGE = datetime.timedelta(hours=1, minutes=1)
+
+
+_filename_re = re.compile(r'^\s*filename=("?)(.+?)\1\s*$')
+
+
+def _filename_from_content_disposition(content_disposition):
+    return next(
+        (match.group(2) for match in (_filename_re.match(sect) for sect in content_disposition.split(";")) if match),
+        None,
+    )
+
+
+def _tag_value_from_tag_set(tag_set, tag_key):
+    return next((tag["Value"] for tag in tag_set if tag["Key"] == tag_key), None)
+
+
+def _tag_set_updated_with_value(tag_set, tag_key, tag_value):
+    return list(chain(
+        (kv for kv in tag_set if kv["Key"] != tag_key),
+        ({"Key": tag_key, "Value": tag_value},),
+    ))
+
+
+class UnknownClamdError(Exception):
+    pass
+
+
+@callbacks.route("/sns/s3/uploaded", methods=['POST'])
+def handle_s3_sns():
+    body_dict = _get_request_body_json()
+
+    # check SNS signature authenticity
+    try:
+        validatesns.validate(
+            body_dict,
+            get_certificate=_get_certificate,
+            certificate_url_regex=VALIDATION_CERTIFICATE_URL_REGEX,
+            max_age=VALIDATION_MAX_AGE,
+        )
+    except (validatesns.ValidationError, requests.exceptions.RequestException) as e:
+        current_app.logger.warning("SNS request body failed signature validation: {validation_error}", extra={
+            "validation_error": e,
+        })
+        abort(400, "SNS request body failed signature validation")
+
+    supported_topic_name = f"s3_file_upload_notification_{current_app.config['DM_ENVIRONMENT']}"
+
+    if body_dict["Type"] == "SubscriptionConfirmation":
```

Comment:
Any chance of a key error here? Will there definitely be a 'Type' key?

Reply:
:man_shrugging: if there ain't, maybe we *should* error and know about it?

Reply:
Yep, fair enough